Handle query commands on demand in main loop

joelim-work opened the pull request:
One limitation of `shell` (`$`) and `shell-wait` (`!`) commands is that they run synchronously on the main loop, so it is not possible to do anything else that requires processing from the main loop. For example commands invoked via `lf -remote "send $id <command>"` will not be processed until the shell command finishes.

This is problematic for `lf -remote "query $id <query>"` calls, since they need to access internal state that is owned by the goroutine running the main loop - delegating the query to the main loop will not work as it is stalled when running synchronous shell commands. The current workaround caches the internal state before invoking synchronous shell commands, but this leads to stale cache entries, especially when running `lf -remote "query $id <query>"` from an external source, as mentioned in #2145.

This PR changes synchronous shell commands so that only the UI is suspended, but the main loop continues to run. This allows queries to be resolved on demand instead of retrieving cached content that could potentially be outdated. Some other refactoring is done as well. This is a large change in terms of how synchronous shell commands work fundamentally, so it would be nice to get some feedback for this, especially if this might break something.

CatsDeservePets commented:
Not sure if this has any use case in the first place, but every `on-quit` command that uses a remote command with `%id` now fails as it gets executed after the client already quit. Previously, the client quit after executing.

----
This also seems to break the following `pre-cd` and `on-cd` hooks:
```sh
cmd pre-cd ${{
	echo pre: $PWD
}}

cmd on-cd ${{
	echo on: $PWD
}}
```
1. I started at /Users/home/Downloads/lf_latest/query2
2. I switched to /Users/home
3. I switched to /Users/home/Downloads 

Output (old):
pre: /Users/home/Downloads/lf_latest/query2
on: /Users/home
pre: /Users/home
on: /Users/home/Downloads

Output (new):
pre: /Users/home
pre: /Users/home/Downloads

joelim-work commented:
@CatsDeservePets Thanks for the feedback.

> every `on-quit` command that uses a remote command with `%id` now fails as it gets executed after the client already quit. Previously, the client quit after executing.

So what actually happens is that `on-quit` is run before the `lf` instance deregisters itself from the server: https://github.com/gokcehan/lf/blob/d2e125d264e85b6b0c4384a564546933f6b4fc87/app.go#L79-L89

But if the command is now run asynchronously, `lf` doesn't wait for it to finish before deregistration. Actually this is already an issue with the existing `shell-pipe` and `shell-async` commands.

> This also seems to break the following `pre-cd` and `on-cd` hooks:
> 
> ```shell
> cmd pre-cd ${{
> 	echo pre: $PWD
> }}
> 
> cmd on-cd ${{
> 	echo on: $PWD
> }}
> ```
> 
> 1. I started at /Users/home/Downloads/lf_latest/query2
> 2. I switched to /Users/home
> 3. I switched to /Users/home/Downloads
> 
> Output (old): pre: /Users/home/Downloads/lf_latest/query2 on: /Users/home pre: /Users/home on: /Users/home/Downloads
> 
> Output (new): pre: /Users/home pre: /Users/home/Downloads

Another good point. The `pre-cd` and `on-cd` hooks are run one after another in succession. In the current code where `shell` and `shell-wait` run synchronously there is no issue. But if these are changed to run asynchronously, then `on-cd` will not wait for `pre-cd` to finish running first, so that design is problematic. Even in the current code it doesn't work for `shell-pipe` commands because they are skipped if there is already one running: https://github.com/gokcehan/lf/blob/d2e125d264e85b6b0c4384a564546933f6b4fc87/app.go#L596-L598

---

So given these issues, I think it's better to not change anything for now. Actually I sort of just submitted this PR as a draft because the idea was interesting and I didn't want to have to throw away my work. I will most likely close this at some point, it introduces too many changes in the actual behavior for it to be able to be merged. Thanks for testing it out at the very least.
